File: Kotlin/Algorithm/src/완전 이진 트리.kt
import java.io.BufferedReader
import java.io.InputStreamReader

private lateinit var arr: MutableList<Int>
private lateinit var res: Array<MutableList<Int>>

fun main(args: Array<String>) = with(BufferedReader(InputStreamReader(System.`in`))) {
    val k = readLine().toInt()
    arr = readLine().split(" ").map { it.toInt() }.toMutableList()
    res = Array(k) { mutableListOf(0) }

    fun treeCurcuit(arr: List<Int>, depth: Int) {
        if (arr.size == 1)  {
            res[depth].add(arr[0])
            return
        }

        var length = arr.size
        var mid = length / 2
        res[depth].add(arr[mid])
        treeCurcuit(arr.subList(0,(mid)), depth+1)
        treeCurcuit(arr.subList(mid+1,arr.size), depth+1)
    }
    treeCurcuit(arr, 0)
    var sb = StringBuilder()
    for (ele in res) {
        sb.append("${ele.subList(1, ele.size).joinToString(" ")}\n")
    }
    println(sb)
}

File: python/백준 온라인 저지/2022.02.07/완전 이진 트리.py
import sys

input = sys.stdin.readline
sys.setrecursionlimit(10000)
k = int(input())
arr = list(map(int, input().split()))
res = [[] for _ in range(1, k+1)]

def binary_separation(arr, depth):
    if len(arr) == 1:
        res[depth].extend(arr)
        return
    
    length = len(arr)
    mid = length // 2
    res[depth].append(arr[mid])
    binary_separation(arr[:mid], depth+1)
    binary_separation(arr[mid+1:], depth+1)
    
binary_separation(arr, 0)


for i in range(k):
    if i == 0:
        print(res[i][0])
    else:
        print(*res[i])